### PR TITLE
feat: add helper for setting a page object in an integration test

### PR DESCRIPTION
**Pull request by alexlafroscia**
I've found myself using this helper in our codebase at work pretty frequently, and figured it might be useful to other people as well

The API is intended to mirror that of other QUnit helpers, where you pass the `hooks` object into the function to add setup/teardown logic.

This has been really helpful for cases where we have a page object for a component and want to use it within its integration tests as well as composing it into other page objects for acceptance tests.

### CODE REVIEW

File: addon-test-support/setup-page.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { create } from './create';
+
+/**
+ * Sets up a reference to `this.page` with the correct context set
+ */
+export default function(hooks, page) {
+  hooks.beforeEach(function() {
+    this.page = create(page);
+    this.page.setContext(this);
```
Comment (ro0gr) [inline addon-test-support/setup-page.js]: What is the purpose of `setContext`? I think it's only applicable for the `moduleForIntegration` tests, but that doesn't seem a case here, cause we use use `setupTest` thing?

Comment (ro0gr) [reply addon-test-support/setup-page.js]: I think I've got it. You just avoid this annoying `setContext` thing.
I think it enforces all the pages which are instantiated with this helper to switch to an integration mode which makes it unusable in the other kind of tests.

Comment (alexlafroscia) [reply addon-test-support/setup-page.js]: We're currently only using this helper for integration tests, since acceptance tests don't need the page object to be created and the context set.

I was under the impression that it was required for you to use the `create` and `setContext` APIs when using page objects in integration tests, and this little helper just made the setup easier. Is that not the case?

**Comment by alexlafroscia** [comment]
Could you elaborate more on what this would do?

```javascript
const page = this.lookupPage('some/page/object');
```

Where is it resolving that path relative to?

File: addon-test-support/setup-page.js (hunk 1 above)
Comment (alexlafroscia) [reply addon-test-support/setup-page.js]: `moduleForIntegration` is the "old" version of the `ember-qunit` API -- `setupRenderingTest` is the "new" version of the same context, so using `setContext` on the page object is required, as far as I know.

Comment (ro0gr) [reply addon-test-support/setup-page.js]: Yes, you are right, `setContext` is required to enable page objects within `moduleForIntegration` tests.

However there is a [support of new test helpers](https://github.com/san650/ember-cli-page-object/pull/357) included in the beta of ember-cli-page-object. I you aware of that? Is there any  reason you want to keep using `moduleFor`?

**Comment by ro0gr** [comment]
> Could you elaborate more on what this would do?

I think it could lookup a page object definition using `require`, and create a page object instance from it, smth like:

```js
function lookupPage(path) {
  const def = require.entries[`${appName}/tests/pages/${path}`].module.exports.default;

  return create(def);
}
```

But now, when I've written this down I realize this isn't really related to the `setupTest` thing cause it has nothing to do with a test context. It could be implemented just as a plain function.

I'm not actually sure if there any real benefits of the `lookupPage` approach. It probably should not pollute your PR, sorry for that!

**Comment by alexlafroscia** [comment]
Gotcha. Additionally I wouldn't want to assume that page objects are inside `tests/pages` of the app or addon -- I've been putting them in `addon-test-support` for addons, so that apps that install the addon have the page objects in an easily import-able location that's only part of the `test` output.

**Comment by ro0gr** [comment]
> I wouldn't want to assume that page objects are inside tests/pages

That's interesting. You've changed my mind 😃

File: addon-test-support/setup-page.js (hunk 1 above)
Comment (ro0gr) [reply addon-test-support/setup-page.js]: As far as I understand, this way it's possible to use the latest test-helpers while keep using Ember jquery events.

There few issues with `setContext`:
- It's limited to legacy Ember integration mode only.
- It has to be set and removed for each page object instance.

I think we need a public way to explicitly specify mode the enable, smth like:

```js
import { setupPageObjects } from 'ember-cli-page-object/test-support';
import { EmberJqueryAdapter } from 'ember-cli-page-object/adapters'

module('...', function(hooks)
  setupPageObjects(hooks, {
    adapter: EmberJqueryAdapter
  });
```

Going to write some issues for this soon.

**Comment by bendemboski** [comment]
My 2 cents -- I think the different test modes and how they relate to different page object execution contexts is already confusing enough that we shouldn't add to the confusion with something like this.

If I understand it, this allows you to use the legacy integration test context with QUnit nested modules. I don't think this is actually possible with `moduleForComponent()` tests -- I don't think they can give you access to the `hooks` that you would need to pass into the `setupPage()` method. So in order to use this, you'd have to use it with a `setupRenderingTest()` or `setupApplicationTest()` test, and then this would force `ember-cli-page-object` to use the execution context meant for `moduleForComponent()` tests.

While that would technically work, it would be using jQuery instead of `@ember/test-helpers`' native DOM helpers, moving the user squarely off of the intended/beaten path for Ember testing. So I don't think this is something we should add support for, because it's supporting an unintended/edge case, and is likely to add to the already-somewhat-problematic confusion about test styles and execution contexts.

**Comment by alexlafroscia** [comment]
> I think the different test modes and how they relate to different page object execution contexts is already confusing enough that we shouldn't add to the confusion with something like this.

Totally agree.

I'll admit that I don't really know anything about execution contexts or when you need jQuery vs. not or why any of that is necessary. There is a really high likelihood that we're using this helper in our tests right now but should actually be doing something else.

**Comment by bendemboski** [comment]
@alexlafroscia the TL;DR is that you want to call `page.setContext(this)` if and only if you're running a `moduleForComponent()` test. If you're running a `moduleForAcceptance()` test, a `setupRenderingTest()` test, or a `setupApplicationTest()`, e-c-p-o will detect it and do the right thing.

Calling `page.setContext(this)` forces e-c-p-o to use synchronous jQuery calls to interact with the DOM. If you're running a `setup*Test()` test, you don't really want that, you want it to auto-detect and use `@ember/test-helpers` helpers with native events and whatnot.

**Comment by alexlafroscia** [comment]
Cool. In that case, I'm going to go and close this PR because it doesn't seem necessary. I'll look at removing this helper in my own app since we don't need it there, either!
